File: 3933-minimum-jumps-to-reach-end-via-prime-teleportation/minimum-jumps-to-reach-end-via-prime-teleportation.py
class Solution:
    def minJumps(self, nums: List[int]) -> int:
        n=len(nums)
        if n == 1:
            return 0

        mx=max(nums)
        prime=list(range(mx+1))
        for i in range(2, int(mx**0.5)+1):
            if prime[i] == i:
                for j in range(i*i, mx+1, i):
                    prime[j]=i
                
        mp=defaultdict(list)
        for i in range(n):
            x=nums[i]
            st=set()
            while x>1:
                t=prime[x]
                if t not in st:
                    mp[t].append(i)
                    st.add(t)
                x//=t

        visited=[False]*n
        q=deque()
        visited[0]=True
        q.append((0, 0))
        while q:
            idx, steps=q.popleft()
            if idx == n-1:
                return steps

            if idx-1>=0 and not visited[idx-1]:
                q.append((idx-1, steps+1))
                visited[idx-1]=True
            if idx+1<n and not visited[idx+1]:
                q.append((idx+1, steps+1))
                visited[idx+1]=True

            if nums[idx]>1 and prime[nums[idx]] == nums[idx]:
                for it in mp[nums[idx]]:
                    if not visited[it]:
                        q.append((it, steps+1))
                        visited[it]=True
                mp[nums[idx]].clear()

        return -1

File: Medium
Medium/3933-minimum-jumps-to-reach-end-via-prime-teleportation/minimum-jumps-to-reach-end-via-prime-teleportation.java
class Solution{
    public int minJumps(int[] nums){
        int n=nums.length;
        if(n == 1)
            return 0;

        int mx=0;
        for(int x: nums){
            mx=Math.max(mx, x);
        }
        int[] prime=new int[mx+1];
        for(int i=0; i<=mx; i++){
            prime[i]=i;
        }

        for(int i=2; i*i<=mx; i++){
            if(prime[i] == i){
                for(int j=i*i; j<=mx; j+=i){
                    prime[j]=i;
                }
            }
        }

        HashMap<Integer, List<Integer>> mp=new HashMap<>();
        for(int i=0; i<n; i++){
            int x=nums[i];
            HashSet<Integer> st=new HashSet<>();
            while(x>1){
                int t=prime[x];
                if(!st.contains(t)){
                    mp.computeIfAbsent(t, k->new ArrayList<>()).add(i);
                    st.add(t);
                }
                x/=t;
            }
        }

        boolean[] visited=new boolean[n];
        Queue<int[]> q=new LinkedList<>();
        visited[0]=true;
        q.offer(new int[]{0, 0});
        while(!q.isEmpty()){
            int[] curr=q.poll();
            int idx=curr[0];
            int steps=curr[1];

            if(idx == n-1)
                return steps;

            if(idx-1>=0 && visited[idx-1] == false){
                q.offer(new int[]{idx-1, steps+1});
                visited[idx-1]=true;
            }
            if(idx+1<n && visited[idx+1] == false){
                q.offer(new int[]{idx+1, steps+1});
                visited[idx+1]=true;
            }

            if(nums[idx]>1 && prime[nums[idx]] == nums[idx]){
                for(int it: mp.get(nums[idx])){
                    if(visited[it] == false){
                        q.offer(new int[]{it, steps+1});
                        visited[it]=true;
                    }
                }
                mp.get(nums[idx]).clear();
            }
        }

        return -1;
    }
}